pkg/controllers/updaterun/controller.go

```diff
@@ -69,7 +84,39 @@ func (r *Reconciler) Reconcile(ctx context.Context, req runtime.Request) (runtim
 		return runtime.Result{}, err
 	}
 
-	// TODO(wantjian): reconcile the clusterStagedUpdateRun.
+	var updatingStageIndex int
+	var toBeUpdatedBindings, toBeDeletedBindings []*placementv1beta1.ClusterResourceBinding
+	var err error
+	initCond := meta.FindStatusCondition(updateRun.Status.Conditions, string(placementv1alpha1.StagedUpdateRunConditionInitialized))
+	if !condition.IsConditionStatusTrue(initCond, updateRun.Generation) {
+		if condition.IsConditionStatusFalse(initCond, updateRun.Generation) {
+			klog.V(2).InfoS("The clusterStagedUpdateRun has failed to initialize", "errorMsg", initCond.Message, "clusterStagedUpdateRun", runObjRef)
+			return runtime.Result{}, nil
+		}
+		if toBeUpdatedBindings, toBeDeletedBindings, err = r.initialize(ctx, &updateRun); err != nil {
+			klog.ErrorS(err, "Failed to initialize the clusterStagedUpdateRun", "clusterStagedUpdateRun", runObjRef)
+			// errInitializedFailed cannot be retried.
+			if errors.Is(err, errInitializedFailed) {
```

Comment:
we also should not retry if the error is "unexpected"

Reply:
I believe I wrapped all terminating errors including unexpected ones as `errInitializedFailed`. I think adding another case would make things more complicated.

Reply:
oh, I miss read the code. However, since the outer error is always failedToInitialize, can we unwrap when we return the error. It might be easier to classify the error and create alert/monitor that way, especially now we emit error types like user error, unexpected or API error.